### PR TITLE
replace owncloud/davclient.js with webdav

### PR DESCRIPTION
**Pull request by dpakach**
### Description
replace owncloud/davclient.js with webdav
Related Issue https://github.com/owncloud/owncloud-sdk/issues/712
Related https://github.com/owncloud/owncloud-sdk/issues/705

This PR replaces `owncloud/davclient.js` with `perry-mitchell/webdav-client`. This makes owncloud-sdk compatible with nodejs.

Changes -
- [x] - Import the code related to making dav request and parsing dav response into new Dav class from https://github.com/owncloud/davclient.js/blob/master/lib/client.js
- [x] - Replace the use of https://github.com/owncloud/davclient.js with the newly created Dav class
- [x] - Unskip and adjust all the tests so that they pass

### CODE REVIEW

**Comment by dpakach** [comment]
### TODO
- [ ] - It Still doesn't work with web fully, make all the tests pass on web (https://github.com/owncloud/web/pull/4738)
          - this will probably require code changes on web
  In web folders are redered as files, this may be because the `d:resoucetype` property used to be an array of xmlNode (eg. [d:collection] where `d:collection` is xmlNode), but it now returns it as array of string (eg. ["d:collection"]) this needs to be handled in web. There may also be other errors that needs to be checked.

Screenshot from 2021-02-17 12-45-57

**Comment by dpakach** [comment]
@LukasHirt Looks like everything is working now. I tested it on web https://github.com/owncloud/web/pull/4738
All tests are passing on web. I mostly just copied the code from owncloud/davclient.js you may find some improvements there.
can you review when you have time?

File: src/dav.js
Comment (individual-it) [inline src/dav.js:55]: what is a sting is passed but something useless like `very deep`?
or it being an empty sting?

File: src/dav.js
Comment (individual-it) [inline src/dav.js:51]: why not having the depth & headers parameter having default values?  
`propFind (path, properties, depth = '0', headers = {}, options = {}) {`

File: src/filesTrash.js
Comment (individual-it) [inline src/filesTrash.js:43]: long line

File: src/filesTrash.js
Comment (individual-it) [inline src/filesTrash.js:68]: long line

File: src/filesTrash.js
Comment (individual-it) [inline src/filesTrash.js:98]: long line

File: src/systemTags.js
Comment (individual-it) [inline src/systemTags.js:45]: long line

File: tests/filesTest.js
Comment (individual-it) [inline tests/filesTest.js:858]: isn't there a variable for that?

File: src/dav.js
Comment (individual-it) [inline src/dav.js:365]: unit tests for this function would be useful to make sure the regex is correct and stays correct

File: src/dav.js
Comment (individual-it) [inline src/dav.js:187]: unit tests for this function would be great, to make sure the composed body is correct.

File: src/dav.js
Comment (individual-it) [inline src/dav.js:152]: unit tests for this function would be great, to make sure the composed body is correct.

File: src/dav.js
Comment (individual-it) [inline src/dav.js:49]: unit tests for this function would be great, to make sure the composed body is correct.

**Review by individual-it — CHANGES_REQUESTED** [review]
wow, what an epic work :rocket: . I'm happy that soo many more tests can be enabled

it would be also interesting to run the acceptance tests of https://github.com/owncloud/web/ with this changes before merging them to make sure it also works correctly in the browser evn.

File: src/dav.js
Comment (dpakach) [reply src/dav.js:187]: I think that is covered very well by the pact tests. These methods make HTTP requests if we want to do unit tests we will need to intercept the HTTP request and check but not really necessary IMO

File: src/dav.js
Comment (dpakach) [reply src/dav.js:55]: If you pass invalid depth then it will be ignored by the server, so it's ok IMO

File: src/dav.js
Comment (kulmann) [inline src/dav.js:70]:
```suggestion
      if (!Object.prototype.hasOwnProperty.call(properties, ii)) {
```
if you do it like this you can get rid of the eslint disable command

File: src/dav.js
Comment (kulmann) [inline src/dav.js:86]: You could assign the depth specific value to a variable and then have simpler return handling:
```js
const body = (depth === '0' ? result.body : result.body[0])
return {
             status: result.status,
             body,
             res: result
}
```

File: src/dav.js
Comment (kulmann) [inline src/dav.js:115]: See the other hasOwnProperty related comment to make eslint happy.

File: src/dav.js
Comment (kulmann) [inline src/dav.js:56]: You had a constant for the tests with `'application/xml; charset=utf-8'` as content-type headers. Could you move that to another location and reuse it in this class as well (there are other occurrences in this class).

File: tests/pactHelper.js
Comment (kulmann) [inline tests/pactHelper.js:111]: Do I get it right, that you could remove the versioning here entirely? Looks like a leftover from the development of this PR. Matchers below are identical.

**Review by kulmann — CHANGES_REQUESTED** [review]

**Comment by dpakach** [comment]
@kulmann I've fixed everything can you review it. Also all tests pass on web https://github.com/owncloud/web/pull/4738

**Review by kulmann — APPROVED** [review]
LGTM now. Impressive piece of work! 🚀

**Comment by dpakach** [comment]
Thanks a lot @kulmann, I've rebased, will merge when CI is green

**Review by individual-it — CHANGES_REQUESTED** [review]
there are now suddenly 81 failures in the provider tests, need investigation
image

**Comment by individual-it** [comment]
I thought that if already passing and approved provider tests would start failing CI would fail, but that seems not to be the case

**Comment by dpakach** [comment]
@individual-it I have confirmed that all the failures **on oc10 provider tests** exist on the previously skipped tests.
There was some failure on public link tests which i have fixed on https://github.com/owncloud/owncloud-sdk/pull/735/commits/c024777722794bd004adfb4e368e15004adfe38b
They are now down to 73 warnings. Can you review and merge now?

**Review by individual-it — APPROVED** [review]

**Pull request merged**
